src/grain/distribution.js

```diff
@@ -0,0 +1,232 @@
+ * cred score)
+ *
+ * In all cases, the timestamp for the distribution is used to determine which cred
+ * scores are in scope. For example, if you create a immediate distribution with a
+ * timestamp in the past, it will reward people with cred in the past time
+ * period, not the current one.
+ */
+import {sum} from "d3-array";
+import {mapToArray} from "../util/map";
+import {type NodeAddressT} from "../core/graph";
+import {type Grain, multiplyFloat, toFloatRatio, ZERO} from "./grain";
+
+export const DISTRIBUTION_VERSION_1 = 1;
+
+export type DistributionStrategy = ImmediateV1 | LifetimeV1;
+
+export type ImmediateV1 = {|
+  +type: "IMMEDIATE",
+  +version: number,
+  +amount: Grain,
+|};
+
+export type LifetimeV1 = {|
+  +type: "LIFETIME",
+  +version: number,
+  +amount: Grain,
+|};
+
+export type GrainReceipt = {|+address: NodeAddressT, amount: Grain|};
+
+export type DistributionV1 = {|
+  +type: "DISTRIBUTION",
+  +timestampMs: number,
+  +version: number,
+  +strategy: DistributionStrategy,
+  +receipts: $ReadOnlyArray<GrainReceipt>,
+|};
+
+export type CredTimeSlice = {|
+  +intervalEndMs: number,
+  +cred: Map<NodeAddressT, number>,
+|};
+
+export type CredHistory = $ReadOnlyArray<CredTimeSlice>;
+
+/**
+ * Compute a full Distribution given:
+ * - the strategy we're using
+ * - the full cred history for all users
+ * - the lifetime earnings of all users
+ * - the timestamp for the distribution
+ */
+export function distribution(
+  strategy: DistributionStrategy,
+  credHistory: CredHistory,
+  earnings: Map<NodeAddressT, Grain>,
+  timestampMs: number
+): DistributionV1 {
+  const filteredSlices = credHistory.filter(
+    (s) => s.intervalEndMs <= timestampMs
+  );
+  if (!filteredSlices.length) {
+    return {
+      type: "DISTRIBUTION",
+      timestampMs,
+      version: DISTRIBUTION_VERSION_1,
+      strategy,
+      receipts: [],
+    };
+  }
+
+  const receipts: $ReadOnlyArray<GrainReceipt> = (function () {
+    switch (strategy.type) {
+      case "IMMEDIATE":
+        if (strategy.version !== 1) {
+          throw new Error(
+            `Unsupported IMMEDIATE strategy: ${strategy.version}`
+          );
+        }
+        const lastSlice = filteredSlices[filteredSlices.length - 1];
+        return computeImmediateReceipts(strategy.amount, lastSlice.cred);
+      case "LIFETIME":
+        if (strategy.version !== 1) {
+          throw new Error(`Unsupported LIFETIME strategy: ${strategy.version}`);
+        }
+        const totalCred = new Map();
+        for (const {cred} of filteredSlices) {
+          for (const [address, ownCred] of cred.entries()) {
+            const existingCred = totalCred.get(address) || 0;
+            totalCred.set(address, existingCred + ownCred);
+          }
+        }
+        return computeLifetimeReceipts(strategy.amount, totalCred, earnings);
+      default:
+        throw new Error(`Unexpected type ${(strategy.type: empty)}`);
+    }
+  })();
```

Comment:
I think we wanted to keep it inline to make it clear that its not meant to be re-used anywhere else and its explicitly internal to the distribution function. However, I do agree that it looks a bit "off", and having the declared function might be a bit more readable for some people, but then it might give the idea that it should be moved outside instead of declared inline. I think if we end up reusing it elsewhere we can move it outside of the function, but for now the IIFE is fine IMO

Reply:
Sorry that was a bit of an unfinished thought.

I'm not opposed using an IIFE :+1:
I'm sharing alternatives because when code has that "bit off" look, exploring different styles sometimes helps in removing accidental complexity.

For example for sake of exploring I might start off like: "what would it take to replace this switch assignment with multiple ternary operations?"

```js
// Some set of named arguments for the underlying computes,
// which we may only need a subset of.
const opts = {/* ... */};
const receipts =
  strat === "IMMEDIATE" ? computeImmediate(opts) :
  strat === "LIFETIME" ? computeLifetime(opts) :
  unknownStratError(strat);
```

Yeah that's not too great, needing functions just to throw. And prettify will butcher this.
But being forced to make new functions to replace the case bodies, and taking named arguments seems nice. What about putting that back into a switch IIFE?

```js
const opts = {/* ... */};
const receipts = (() => {
  switch (strat) {
    case "IMMEDIATE":
      return computeImmediate(opts);
    case "LIFETIME":
      return computeLifetime(opts);
    default:
      throw new Error(`Unexpected type ${(strat: empty)}`);
  }
})();
```

That seems not bad at all. Maybe the cases just need to be simplified like this.
Then, what would the named function vs IIFE look like?

```js
function _computeReceipts(strat, opts) {
  switch (strat) {
    case "IMMEDIATE":
      return computeImmediate(opts);
    case "LIFETIME":
      return computeLifetime(opts);
    default:
      throw new Error(`Unexpected type ${(strat: empty)}`);
  }
}

const opts = {/* ... */};
return {
  ...rest,
  receipts: _computeReceipts(strat, opts),
};
```

I think this looks nice in a "nothing special happening here" kind of way.
Maybe it's worth using some of this.

---

Other note: If you'd like to indicate it's not intended for use outside this file. We're using a convention where occasionally we export functions with a `_` prefix in the name. Indicating they're "private" functions, but we might want to export them anyway for unit testing.